From 0e1425c06977ed59fd1568c648835df041e8126e Mon Sep 17 00:00:00 2001
From: Michal Kowalczyk <michal.kowalczyk@plume.com>
Date: Tue, 15 Oct 2019 14:20:07 +0200
Subject: [PATCH 2/3] inet: inject br-wan and br-lan handling

As the OSN API doesn't allow to override L2 stuff we cannot fully
utilize it in case of RDK platform which has couple of specific
interfaces. Those interfaces resides on other CPU so we cannot access
them directly.
Those interfaces are accessed using devicenfo.sh script.
---
 src/lib/inet/src/inet_eth.c | 104 +++++++++++++++++++++++++++++++++++-
 1 file changed, 102 insertions(+), 2 deletions(-)

diff --git a/src/lib/inet/src/inet_eth.c b/src/lib/inet/src/inet_eth.c
index 47db5e6c..fd9eb854 100644
--- a/src/lib/inet/src/inet_eth.c
+++ b/src/lib/inet/src/inet_eth.c
@@ -13,6 +13,7 @@
 #include "inet.h"
 #include "inet_base.h"
 #include "inet_eth.h"
+#include "osync_hal.h"
 
 #include "execsh.h"
 
@@ -449,6 +450,14 @@ bool ifreq_ioctl(const char *ifname, int cmd, struct ifreq *req)
     return true;
 }
 
+/** On RDK some interfaces are on different CPU.
+ *  We must treat them separetely.
+ */
+static bool remote_interface(const char *ifname)
+{
+    return (!strcmp(ifname, "br-home") || !strcmp(ifname, "br-wan"));
+}
+
 /**
  * Check if the interface exists
  */
@@ -456,6 +465,15 @@ bool ifreq_exists(const char *ifname, bool *exists)
 {
     struct ifreq ifr;
 
+    if (remote_interface(ifname))
+    {
+        /* We don't have control over that interfaces.
+         * We must assume it always exists.
+         */
+        *exists = true;
+        return true;
+    }
+
     /* First get the current flags */
     if (!ifreq_ioctl(ifname, SIOCGIFINDEX, &ifr))
     {
@@ -477,6 +495,11 @@ bool ifreq_status_set(const char *ifname, bool up)
 {
     struct ifreq ifr;
 
+    if (remote_interface(ifname))
+    {
+        return true;
+    }
+
     /* First get the current flags */
     if (!ifreq_ioctl(ifname, SIOCGIFFLAGS, &ifr))
     {
@@ -516,6 +539,12 @@ bool ifreq_status_get(const char *ifname, bool *up)
 {
     struct ifreq ifr;
 
+    if (remote_interface(ifname))
+    {
+        *up = true;
+        return true;
+    }
+
     /* First get the current flags */
     if (!ifreq_ioctl(ifname, SIOCGIFFLAGS, &ifr))
     {
@@ -538,6 +567,12 @@ bool ifreq_running_get(const char *ifname, bool *up)
 {
     struct ifreq ifr;
 
+    if (remote_interface(ifname))
+    {
+        *up = true;
+        return true;
+    }
+
     /* First get the current flags */
     if (!ifreq_ioctl(ifname, SIOCGIFFLAGS, &ifr))
     {
@@ -559,6 +594,11 @@ bool ifreq_ipaddr_set(const char *ifname, osn_ip_addr_t ipaddr)
 {
     struct ifreq ifr;
 
+    if (remote_interface(ifname))
+    {
+        return true;
+    }
+
     ifr.ifr_addr.sa_family = AF_INET;
     memcpy(&((struct sockaddr_in *)&ifr.ifr_addr)->sin_addr.s_addr, &ipaddr, sizeof(ipaddr));
     if (!ifreq_ioctl(ifname, SIOCSIFADDR, &ifr))
@@ -579,9 +619,23 @@ bool ifreq_ipaddr_set(const char *ifname, osn_ip_addr_t ipaddr)
 bool ifreq_ipaddr_get(const char *ifname, osn_ip_addr_t *ipaddr)
 {
     struct ifreq ifr;
+    osync_hal_inet_iface_config_t iface_config;
 
     *ipaddr = OSN_IP_ADDR_INIT;
 
+    if (remote_interface(ifname))
+    {
+        memset(&iface_config, 0, sizeof(iface_config));
+        if (osync_hal_inet_get_iface_config(ifname, &iface_config)
+                != OSYNC_HAL_SUCCESS)
+        {
+            LOG(ERR, "RDK: cannot get iface config for %s", ifname);
+            return false;
+        }
+
+        return osn_ip_addr_from_str(ipaddr, iface_config.inet_addr);
+    }
+
     ifr.ifr_addr.sa_family = AF_INET;
 
     *ipaddr = OSN_IP_ADDR_INIT;
@@ -603,6 +657,11 @@ bool ifreq_netmask_set(const char *ifname, osn_ip_addr_t netmask)
 {
     struct ifreq ifr;
 
+    if (remote_interface(ifname))
+    {
+        return true;
+    }
+
     ifr.ifr_addr.sa_family = AF_INET;
     ((struct sockaddr_in *)&ifr.ifr_addr)->sin_addr = netmask.ia_addr;
     if (!ifreq_ioctl(ifname, SIOCSIFNETMASK, &ifr))
@@ -624,9 +683,14 @@ bool ifreq_netmask_get(const char *ifname, osn_ip_addr_t *netmask)
 {
     struct ifreq ifr;
 
+    *netmask = OSN_IP_ADDR_INIT;
+    if (remote_interface(ifname))
+    {
+        return true;
+    }
+
     ifr.ifr_addr.sa_family = AF_INET;
 
-    *netmask = OSN_IP_ADDR_INIT;
     if (!ifreq_ioctl(ifname, SIOCGIFNETMASK, &ifr))
     {
         return false;
@@ -648,6 +712,11 @@ bool ifreq_bcaddr_set(const char *ifname, osn_ip_addr_t bcaddr)
 {
     struct ifreq ifr;
 
+    if (remote_interface(ifname))
+    {
+        return true;
+    }
+
     ifr.ifr_addr.sa_family = AF_INET;
     ((struct sockaddr_in *)&ifr.ifr_addr)->sin_addr = bcaddr.ia_addr;
     if (!ifreq_ioctl(ifname, SIOCSIFBRDADDR, &ifr))
@@ -671,6 +740,11 @@ bool ifreq_bcaddr_get(const char *ifname, osn_ip_addr_t *bcaddr)
 
     *bcaddr = OSN_IP_ADDR_INIT;
 
+    if (remote_interface(ifname))
+    {
+        return true;
+    }
+
     ifr.ifr_addr.sa_family = AF_INET;
 
     if (!ifreq_ioctl(ifname, SIOCGIFBRDADDR, &ifr))
@@ -693,6 +767,11 @@ bool ifreq_mtu_set(const char *ifname, int mtu)
 {
     struct ifreq ifr;
 
+    if (remote_interface(ifname))
+    {
+        return true;
+    }
+
     ifr.ifr_mtu = mtu;
 
     if (!ifreq_ioctl(ifname, SIOCSIFMTU, &ifr))
@@ -714,6 +793,12 @@ bool ifreq_mtu_get(const char *ifname, int *mtu)
 {
     struct ifreq ifr;
 
+    if (remote_interface(ifname))
+    {
+        *mtu = 1500;
+        return true;
+    }
+
     if (!ifreq_ioctl(ifname, SIOCGIFMTU, &ifr))
     {
         LOG(ERR, "inet_eth: %s: SIOCGIFMTU failed. Error retrieving the MTU: %s",
@@ -735,6 +820,22 @@ bool ifreq_mtu_get(const char *ifname, int *mtu)
 bool ifreq_hwaddr_get(const char *ifname, osn_mac_addr_t *macaddr)
 {
     struct ifreq ifr;
+    osync_hal_inet_iface_config_t iface_config;
+
+    *macaddr = OSN_MAC_ADDR_INIT;
+
+    if (remote_interface(ifname))
+    {
+        memset(&iface_config, 0, sizeof(iface_config));
+        if (osync_hal_inet_get_iface_config(ifname, &iface_config)
+                != OSYNC_HAL_SUCCESS)
+        {
+            LOG(ERR, "RDK: cannot get iface config for %s", ifname);
+            return false;
+        }
+
+        return osn_mac_addr_from_str(macaddr, iface_config.mac_str);
+    }
 
     /* Get the MAC(hardware) address */
     if (!ifreq_ioctl(ifname, SIOCGIFHWADDR, &ifr))
@@ -746,7 +847,6 @@ bool ifreq_hwaddr_get(const char *ifname, osn_mac_addr_t *macaddr)
         return false;
     }
 
-    *macaddr = OSN_MAC_ADDR_INIT;
     /* Copy the address */
     memcpy(macaddr->ma_addr, ifr.ifr_addr.sa_data, sizeof(macaddr->ma_addr));
 
-- 
2.17.1

